packages/node/src/methods/app-instance/take-action/operation.ts

```diff
@@ -0,0 +1,41 @@
+import { AppInstance } from "@counterfactual/machine";
+import { AppState } from "@counterfactual/types";
+import { Contract } from "ethers";
+import { Provider } from "ethers/providers";
+
+import { ERRORS } from "../../errors";
+
+export async function generateNewAppInstanceState(
+  appInstance: AppInstance,
+  action: any,
+  provider: Provider
+): Promise<AppState> {
+  if (
+    !appInstance.appInterface.addr ||
+    appInstance.appInterface.addr.trim() === ""
+  ) {
+    return Promise.reject(ERRORS.NO_APP_CONTRACT_ADDR);
+  }
+
+  const abi = [
+    `function applyAction(${appInstance.appInterface.stateEncoding}, ${
+      appInstance.appInterface.actionEncoding
+    }) pure returns (bytes)`
```

Comment:
Two things:

1. Probably it would be nice to put this string construction into a file called `function-signatures.ts` or something like that and have it be called e.g., `getApplyActionSignature(stateEncoding, actionEncoding)`.

2. Inside that function you can use newlines, e.g.,

```typescript
const signature = `
    function applyAction(
        ${appInstance.appInterface.stateEncoding} state,
        ${appInstance.appInterface.actionEncoding} action
    )
        pure
        returns (bytes)
`;
```

3. Furthermore, this is actually in flux at the moment. Very soon everything will use the exact same interface:

`function applyAction(bytes, bytes) pure returns (bytes)`

Reply:
I tried `function applyAction(bytes, bytes) pure returns (bytes)`

It failed.

Reply:
Whenever that ends up working we can update.